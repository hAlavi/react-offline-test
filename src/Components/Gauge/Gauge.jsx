import React, { useState, useEffect } from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';

const GaugeWrapper = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  line-height: 1;
  font-weight: 300;
  font-size: 1.5em;
  font-family: 'Lato', sans-serif;
  user-select: none;
  margin: 10px;
  background-color: #ffffff;
`;

const GaugeModal = styled.div`
  margin: 10px;
  border: 1px solid #ffffff;
  border-radius: 3px;
  background-color: #ffffff;
`;

const StyledProgress = styled.div`
  position: relative;
  display: inline-block;
`;

const StyledCircle = styled.svg`
  transform: rotate(-90deg);
  fill: none;
  stroke-linecap: round;
`;
const StyledCircleBackGround = styled.circle`
  stroke: #e6e6e6;
`;
const StyledCircleFill = styled.circle`
  stroke: #65c3de;
`;

const PercentWrapper = styled.div``;
const PercentContent = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  height: 100%;
`;

const ControlInput = styled.input`
  margin-top: 4px;
`;
const CircleGauge = ({ size, progress }) => {
  const center = size / 2;
  const strokeWidth = size * 0.1;
  const radius = size / 2 - strokeWidth / 2;
  const circumference = 2 * Math.PI * radius;
  const offset = ((100 - progress) / 100) * circumference;

  const style = {
    strokeDashoffset: offset,
  };

  return (
    <StyledCircle width={size} height={size} viewBox={`0 0 ${size} ${size}`}>
      <StyledCircleBackGround
        cx={center}
        cy={center}
        r={radius}
        strokeWidth={strokeWidth}
      />
      <StyledCircleFill
        style={style}
        cx={center}
        cy={center}
        r={radius}
        strokeWidth={strokeWidth}
        strokeDasharray={circumference}
      />
    </StyledCircle>
  );
};
CircleGauge.propTypes = {
  size: PropTypes.number,
  progress: PropTypes.number.isRequired,
};

CircleGauge.defaultProps = { size: 120 };

const PercentText = ({ progress }) => {
  return (
    <PercentWrapper>
      <PercentContent>{progress}%</PercentContent>
    </PercentWrapper>
  );
};

PercentText.propTypes = {
  progress: PropTypes.number.isRequired,
};

const ProgressWrapper = ({ progress }) => {
  return (
    <StyledProgress>
      <CircleGauge progress={progress} />
      <PercentText progress={progress} />
    </StyledProgress>
  );
};

ProgressWrapper.propTypes = {
  progress: PropTypes.number.isRequired,
};

const Gauge = ({ value, testId }) => {
  const [progress, setProgress] = useState(0);
  useEffect(() => {
    const timerID = setInterval(() => {
      const delta = progress + 1;
      if (delta <= value) {
        setProgress(delta);
      } else {
        clearInterval(timerID);
      }
    }, 10);
    return function cleanup() {
      clearInterval(timerID);
    };
  });

  return (
    <GaugeWrapper data-testid={testId}>
      <GaugeModal data-testid={`${testId}-modal`}>
        <ProgressWrapper progress={progress} />
      </GaugeModal>
      <ControlInput
        data-testid={`${testId}-input`}
        type="range"
        value={progress}
        min={0}
        max={100}
        readOnly
      />
    </GaugeWrapper>
  );
};

Gauge.propTypes = {
  value: PropTypes.number,
  testId: PropTypes.string,
};

Gauge.defaultProps = {
  value: 60,
  testId: 'gauge',
};

export default Gauge;
