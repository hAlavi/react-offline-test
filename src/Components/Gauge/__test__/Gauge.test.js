import React from 'react';
import { render, cleanup, waitForElement, act } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import Gauge from '../Gauge';
import 'jest-styled-components';

afterEach(cleanup);

describe('Unit tests for the Gauge Component', () => {
  describe('Visual Snapshot testing', () => {
    it('Renders without crashing', () => {
      const { container } = render(<Gauge value={10} />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('Functional unit tests', () => {
    it('it should have specific parts', () => {
      const { getByTestId } = render(<Gauge testId="gauge-test" value={10} />);

      expect(getByTestId('gauge-test')).toBeTruthy();
      expect(getByTestId('gauge-test-modal')).toBeTruthy();
      expect(getByTestId('gauge-test-input')).toBeTruthy();
    });

    it('it should call timers to update the internal state', async () => {
      jest.useFakeTimers();
      const setState = jest.fn();
      const useStateSpy = jest.spyOn(React, 'useState');
      useStateSpy.mockImplementation(init => [init, setState]);

      const container = render(<Gauge testId="gauge-test" value={10} />);
      const { getByTestId } = container;

      //
      expect(getByTestId('gauge-test')).toBeTruthy();
      expect(setInterval).toHaveBeenCalled();
    });

    it('it should have the value after reached the internal state equal to value', async () => {
      await act(async () => {
        const { getByTestId, getByText } = render(
          <Gauge testId="gauge-test" value={10} />,
        );

        expect(getByTestId('gauge-test')).toBeTruthy();
        jest.advanceTimersByTime(1000);
        //
        await waitForElement(() => getByText('10%'));
        expect(getByText('10%')).toBeTruthy();
      });
    });
  });
});
