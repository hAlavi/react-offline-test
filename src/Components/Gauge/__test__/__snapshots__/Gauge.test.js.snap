// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit tests for the Gauge Component Visual Snapshot testing Renders without crashing 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  line-height: 1;
  font-weight: 300;
  font-size: 1.5em;
  font-family: 'Lato',sans-serif;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  margin: 10px;
  background-color: #ffffff;
}

.c1 {
  margin: 10px;
  border: 1px solid #ffffff;
  border-radius: 3px;
  background-color: #ffffff;
}

.c2 {
  position: relative;
  display: inline-block;
}

.c3 {
  -webkit-transform: rotate(-90deg);
  -ms-transform: rotate(-90deg);
  transform: rotate(-90deg);
  fill: none;
  stroke-linecap: round;
}

.c4 {
  stroke: #e6e6e6;
}

.c5 {
  stroke: #65c3de;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 100%;
}

.c7 {
  margin-top: 4px;
}

<div
  class="c0"
  data-testid="gauge"
>
  <div
    class="c1"
    data-testid="gauge-modal"
  >
    <div
      class="c2"
    >
      <svg
        class="c3"
        height="120"
        viewBox="0 0 120 120"
        width="120"
      >
        <circle
          class="c4"
          cx="60"
          cy="60"
          r="54"
          stroke-width="12"
        />
        <circle
          class="c5"
          cx="60"
          cy="60"
          r="54"
          stroke-dasharray="339.29200658769764"
          stroke-width="12"
          style="stroke-dashoffset: 339.29200658769764;"
        />
      </svg>
      <div
        class=""
      >
        <div
          class="c6"
        >
          0
          %
        </div>
      </div>
    </div>
  </div>
  <input
    class="c7"
    data-testid="gauge-input"
    max="100"
    min="0"
    readonly=""
    type="range"
    value="0"
  />
</div>
`;
